### ISSUE TITLE
Crash on multiple buffer table inserts

### ISSUE THREAD

**Issue by genzgd** [ClickHouse/ClickHouse#26382]
ClickHouse 21.4.4

Crash occurred when doing fairly intense inserts (production traffic plus backfill using INSERT INTO) into a buffer table.  There were two virtually identical stack traces, first one below.

Table definition:

```CREATE TABLE comcast_xcr.atsec_15m_bf
(
    `datetime` DateTime,
    `svc_type` String,
    `svc` String,
    `cache_group` String,
    `cache_result` Enum8('hit' = 1, 'miss' = 2, 'error' = 3, 'abort' = 4, 'stale' = 5),
    `client_response` Enum8('none' = 0, 'success' = 2, 'redirect' = 3, 'client_error' = 4, 'server_error' = 5, '502' = 6, '503' = 7),
    `parent_response` Enum8('none' = 0, 'success' = 2, 'redirect' = 3, 'client_error' = 4, 'server_error' = 5, '502' = 6, '503' = 7),
    `event_count` UInt64,
    `served_bytes` UInt64,
    `parent_bytes` UInt64,
    `ttms_avg` AggregateFunction(avg, UInt32),
    `ttms_quants` AggregateFunction(quantilesTiming(0.99, 0.95, 0.9), UInt32),
    `chi_count` AggregateFunction(uniq, FixedString(16)),
    `manifest_count` UInt64,
    `fragment_count` UInt64,
    `ttms_gt_250` UInt64,
    `ttms_gt_750` UInt64,
    `ttms_gt_1000` UInt64,
    `ttms_gt_2000` UInt64,
    `ttms_gt_5000` UInt64,
    `stms_gt_300` UInt64,
    `stms_gt_500` UInt64,
    `stms_gt_1000` UInt64,
    `ttms_gt_1200` UInt64,
    `stms_avg` AggregateFunction(avgIf, Int32, UInt8),
    `stms_quants` AggregateFunction(quantilesTimingIf(0.99, 0.95, 0.9), Int32, UInt8)
)
ENGINE = Buffer('comcast_xcr', 'atsec_15m', 1, 30, 300, 100000, 1000000, 50000000, 200000000) 
```

Crash:

```2021.07.15 17:45:49.101227 [ 2622510 ] {} <Fatal> BaseDaemon: Stack trace: 0xf023c14 0xf025690 0xf1c286b 0xf1bfa15 0xff9cc6e 0xff9bd2b 0xffa55ba 0xffa995a 0xffa8916 0xffa20f3 0xffa2dbd 0xffa15eb 0x1019b9e5 0x1019b5da 0x1034c8ba 0x101d4c7d 0x101d1931 0x101d6476 0x8954fef 0x8958a83 0x7f30657ab14a 0x7f30654dcf23
2021.07.15 17:45:49.101295 [ 2622510 ] {} <Fatal> BaseDaemon: 1. DB::CompressedReadBufferBase::readCompressedData(unsigned long&, unsigned long&, bool) @ 0xf023c14 in /usr/bin/clickhouse
2021.07.15 17:45:49.101305 [ 2622510 ] {} <Fatal> BaseDaemon: 2. DB::CompressedReadBufferFromFile::nextImpl() @ 0xf025690 in /usr/bin/clickhouse
2021.07.15 17:45:49.107677 [ 2622510 ] {} <Fatal> BaseDaemon: 3. DB::SerializationAggregateFunction::deserializeBinaryBulk(DB::IColumn&, DB::ReadBuffer&, unsigned long, double) const @ 0xf1c286b in /usr/bin/clickhouse
2021.07.15 17:45:49.115493 [ 2622510 ] {} <Fatal> BaseDaemon: 4. DB::ISerialization::deserializeBinaryBulkWithMultipleStreams(COW<DB::IColumn>::immutable_ptr<DB::IColumn>&, unsigned long, DB::ISerialization::DeserializeBinaryBulkSettings&, std::__1::shared_ptr<DB::ISerialization::DeserializeBinaryBulkState>&, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, COW<DB::IColumn>::immutable_ptr<DB::IColumn> > > >*) const @ 0xf1bfa15 in /usr/bin/clickhouse
2021.07.15 17:45:49.115975 [ 2622510 ] {} <Fatal> BaseDaemon: 5. DB::MergeTreeReaderWide::readData(DB::NameAndTypePair const&, COW<DB::IColumn>::immutable_ptr<DB::IColumn>&, unsigned long, bool, unsigned long, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, COW<DB::IColumn>::immutable_ptr<DB::IColumn> > > >&) @ 0xff9cc6e in /usr/bin/clickhouse
2021.07.15 17:45:49.115997 [ 2622510 ] {} <Fatal> BaseDaemon: 6. DB::MergeTreeReaderWide::readRows(unsigned long, bool, unsigned long, std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn> > >&) @ 0xff9bd2b in /usr/bin/clickhouse
2021.07.15 17:45:49.116009 [ 2622510 ] {} <Fatal> BaseDaemon: 7. DB::MergeTreeRangeReader::DelayedStream::finalize(std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn> > >&) @ 0xffa55ba in /usr/bin/clickhouse
2021.07.15 17:45:49.116016 [ 2622510 ] {} <Fatal> BaseDaemon: 8. DB::MergeTreeRangeReader::continueReadingChain(DB::MergeTreeRangeReader::ReadResult&, unsigned long&) @ 0xffa995a in /usr/bin/clickhouse
2021.07.15 17:45:49.116024 [ 2622510 ] {} <Fatal> BaseDaemon: 9. DB::MergeTreeRangeReader::read(unsigned long, std::__1::deque<DB::MarkRange, std::__1::allocator<DB::MarkRange> >&) @ 0xffa8916 in /usr/bin/clickhouse
2021.07.15 17:45:49.116041 [ 2622510 ] {} <Fatal> BaseDaemon: 10. DB::MergeTreeBaseSelectProcessor::readFromPartImpl() @ 0xffa20f3 in /usr/bin/clickhouse
2021.07.15 17:45:49.116047 [ 2622510 ] {} <Fatal> BaseDaemon: 11. DB::MergeTreeBaseSelectProcessor::readFromPart() @ 0xffa2dbd in /usr/bin/clickhouse
2021.07.15 17:45:49.116053 [ 2622510 ] {} <Fatal> BaseDaemon: 12. DB::MergeTreeBaseSelectProcessor::generate() @ 0xffa15eb in /usr/bin/clickhouse
2021.07.15 17:45:49.116069 [ 2622510 ] {} <Fatal> BaseDaemon: 13. DB::ISource::tryGenerate() @ 0x1019b9e5 in /usr/bin/clickhouse
2021.07.15 17:45:49.116081 [ 2622510 ] {} <Fatal> BaseDaemon: 14. DB::ISource::work() @ 0x1019b5da in /usr/bin/clickhouse
2021.07.15 17:45:49.116088 [ 2622510 ] {} <Fatal> BaseDaemon: 15. DB::SourceWithProgress::work() @ 0x1034c8ba in /usr/bin/clickhouse
2021.07.15 17:45:49.116094 [ 2622510 ] {} <Fatal> BaseDaemon: 16. ? @ 0x101d4c7d in /usr/bin/clickhouse
2021.07.15 17:45:49.119231 [ 2622510 ] {} <Fatal> BaseDaemon: 17. DB::PipelineExecutor::executeStepImpl(unsigned long, unsigned long, std::__1::atomic<bool>*) @ 0x101d1931 in /usr/bin/clickhouse
2021.07.15 17:45:49.119240 [ 2622510 ] {} <Fatal> BaseDaemon: 18. ? @ 0x101d6476 in /usr/bin/clickhouse
2021.07.15 17:45:49.119250 [ 2622510 ] {} <Fatal> BaseDaemon: 19. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x8954fef in /usr/bin/clickhouse
2021.07.15 17:45:49.119255 [ 2622510 ] {} <Fatal> BaseDaemon: 20. ? @ 0x8958a83 in /usr/bin/clickhouse
2021.07.15 17:45:49.119268 [ 2622510 ] {} <Fatal> BaseDaemon: 21. start_thread @ 0x814a in /usr/lib64/libpthread-2.28.so
2021.07.15 17:45:49.119278 [ 2622510 ] {} <Fatal> BaseDaemon: 22. clone @ 0xfcf23 in /usr/lib64/libc-2.28.so
```

**Comment by alexey-milovidov** [comment]
It's interesting. Looks like the culprit is Buffer + AggregateFunction states.
Is there any chance to construct some synthetic reproducer?

**Comment by alexey-milovidov** [comment]
How often does the issue reproduce?

It will be helpful to temporarily run the build with Address Sanitizer under load (it will run 3x..10x slower) until the issue reproduce and collect the report in /var/log/clickhouse-server/stderr.

The build with ASan is located here:
![Screenshot_20210715_214140](https://user-images.githubusercontent.com/18581488/125840458-316c5936-2aaf-4554-a996-f585854a3554.png)
(go to the details and select `clang` build with sanitizer = address and package type = binary)

https://clickhouse-builds.s3.yandex.net/26367/241bb3757ae15d03820c406de5fa083b2768c967/clickhouse_build_check/clang-11_relwithdebuginfo_address_bundled_unsplitted_disable_False_binary/clickhouse

You can run this binary without installation as:
```
wget https://clickhouse-builds.s3.yandex.net/26367/241bb3757ae15d03820c406de5fa083b2768c967/clickhouse_build_check/clang-11_relwithdebuginfo_address_bundled_unsplitted_disable_False_binary/clickhouse

chmod a+x ./clickhouse

sudo systemctl stop clickhouse
# or sudo service clickhouse stop

sudo -u clickhouse ./clickhouse server --config /etc/clickhouse-server/config.xml --daemon
```

**Comment by genzgd** [comment]
This has not happened before in the week+ we've been using these buffer tables.  It only happened today (once) while running "backfill" queries.  I've just started the major backfills and ~1000 inserts (INSERT INTO from a SELECT on a similar table) across 10 shards worked successfully before the crash.  I've run ~5000 inserts since, but only half the size (30 minutes of raw data instead of 60) and it hasn't happened since reducing the size.  So like most of our production problems, not easy to reproduce.  I'll try the larger insert again on a different table but probably not until next week.

Also I think I missed that actual exception above, sorry:

```
2021.07.15 17:45:49.101113 [ 2622510 ] {} <Fatal> BaseDaemon: ########################################
2021.07.15 17:45:49.101185 [ 2622510 ] {} <Fatal> BaseDaemon: (version 21.4.4.30 (official build), build id: E3FA92117218D182F17C14F864FF4ED3D3689BFE) (from thread 1742699) (query_id: backfill_1626371148892) Received signal Segmentation fault (11)
2021.07.15 17:45:49.101215 [ 2622510 ] {} <Fatal> BaseDaemon: Address: 0x7eb32a6c77cb Access: write. Attempted access has violated the permissions assigned to the memory area.
2021.07.15 17:45:49.101227 [ 2622510 ] {} <Fatal> BaseDaemon: Stack trace: 0xf023c14 0xf025690 0xf1c286b 0xf1bfa15 0xff9cc6e 0xff9bd2b 0xffa55ba 0xffa995a 0xffa8916 0xffa20f3 0xffa2dbd 0xffa15eb 0x1019b9e5 0x1019b5da 0x1034c8ba 0x101d4c7d 0x101d1931 0x101d6476 0x8954fef 0x8958a83 0x7f30657ab14a 0x7f30654dcf23
```

The larger size and the illegal access exception suggest maybe something isn't calculating the size of the aggregateState correctly and extending past the buffer?

**Comment by azat** [comment]
It seems that #24158 is also related (although it is about LowCardinality).
It will be very helpful if you can run version of the server with ASAN as suggested by @alexey-milovidov.

@genzgd And all similar issues reported by you, are you sure that your hardware is fine? Do you have memory with ECC any corrected/uncorrected errors?
By the way have you resolved #24158 ?

**Comment by alexey-milovidov** [comment]
Also I suggest to check:
- does it reproduce if you upgrade to the latest stable version 21.7?
- does it reproduce if you remove the usage of Buffer tables.

**Comment by genzgd** [comment]
Our hardware is fine.  :)  We run close to 200 ClickHouse servers, with 64 or 128 cores and up to 768GB of ECC ram with periodic memory testing, and we replace the DIMMs if ECC is showing an elevated level of errors.  We've only seen no memory related issues or crashes except in connection with buffer tables over the last year or more.

We've continued the backfill queries (probably 100,000 at this point) with smaller batches and it has not reproduced, which makes me think the size of the insert related to the buffer size might be related.  

We have not resolved #24158.  Our workaround has been to change the columns in the affected target tables to use String instead of LowCardinality(String).   As I've said before, the large size of our installation and the small size of my team make it difficult to find time to reproduce this issues that only happen rarely under heavy load.  When we get more caught up (right now we're scaling up for Olympics coverage) we hopefully will be able to revisit that problem.  I appreciate how difficult it is to track down these sorts of things if you can't reliably reproduce, and I apologize for not being able to fully follow up when I file these issues.

**Comment by alexey-milovidov** [comment]
@genzgd The Olympics has ended, I hope there were no issues during coverage?

Maybe there is a time to try to find the reproducing scenario...

**Comment by genzgd** [comment]
Our production clusters ran well during the Olympics coverage, thanks for asking. :)  We also finished major backfills on our aggregate tables with no additional crashes, however we switched to inserting to the file based/destination tables since buffer tables don't really provide any benefit in the scenario of inserting large historical batches which are immediately flushed to disk anyway.  I still believe that this issue (26382) is related to insert sizes that are too big for the buffer tables, and that's relatively easy to avoid, so I don't know that we'll put much priority on recreating this issue if you want to close it.  Of course my first hope when filing a crash incident is that someone will look at the code and say "oh, there's the problem"; I understand if they get closed if reproduction is difficult.

I do intend to go back to trying to reproduce #24158, or at least prove that we can use LowCardinality columns in buffer tables without crashing.  We have also started testing 21.8 and intend to upgrade to that version over the next couple of weeks (I'm OOO quite a bit for the remainder of August).  There are some fixes/improvements around replication queues in recent versions that I'm anxious to deploy.  Once 21.8 has been deployed we will do some testing with LowCardinality columns in buffer tables.
